main.cu

```diff
@@ -174,144 +133,187 @@ __device__ inline int getTreeHeight(int x, int z) {
-                    // yellow flowers
-                    advance_774(rand);
-                    // red flowers
-                    if (random_next(&rand, 1) == 0) {
-                        advance_387(rand);
-                    }
-                    // brown mushroom
-                    if (random_next(&rand, 2) == 0) {
-                        advance_387(rand);
-                    }
-                    // red mushroom
-                    if (random_next(&rand, 3) == 0) {
-                        advance_387(rand);
-                    }
-                    // reeds
-                    advance_830(rand);
-                    // pumpkins
-                    if (random_next(&rand, 5) == 0) {
-                        advance_387(rand);
-                    }
-
-                    for (int i = 0; i < 50; i++) {
-                        bool waterfall_matches = random_next(&rand, 4) == WATERFALL_X;
-                        waterfall_matches &= random_next_int(&rand, random_next_int(&rand, 120) + 8) == WATERFALL_Y;
-                        waterfall_matches &= random_next(&rand, 4) == WATERFALL_Z;
-                        any_population_matches |= waterfall_matches;
-                    }
-                    rand = before_rest;
+            if (treeFlags == ((1 << (OTHER_TREE_COUNT + 1)) - 1)) {
+
+                Random before_rest = rand;
+                // yellow flowers
+                advance_774(rand);
+                // red flowers
+                if (random_next(&rand, 1) == 0) advance_387(rand);
+                // brown mushroom
+                if (random_next(&rand, 2) == 0) advance_387(rand);
+                // red mushroom
+                if (random_next(&rand, 3) == 0) advance_387(rand);
+
+                // reeds
+                advance_830(rand);
+                // pumpkins
+                if (random_next(&rand, 5) == 0) advance_387(rand);
+
+                for (int i = 0; i < 50; i++) {
+                    bool waterfall_matches = random_next(&rand, 4) == WATERFALL_X;
+                    waterfall_matches &= random_next_int(&rand, random_next_int(&rand, 120) + 8) == WATERFALL_Y;
+                    waterfall_matches &= random_next(&rand, 4) == WATERFALL_Z;
+                    any_population_matches |= waterfall_matches;
                 }
+                rand = before_rest;
             }
+        }
 
-            this_res &= any_population_matches;
+        if (any_population_matches) {
+            Random start_chunk_rand = start;
+            advance_m3759(start_chunk_rand);
 
-            if (this_res) {
-                Random start_chunk_rand = start;
-                advance_m3759(start_chunk_rand);
+            int index = atomicAdd(num_seeds, 1);
+            seeds[index] = start_chunk_rand;
+        }
 
-                int index = atomicAdd(num_seeds, 1);
-                seeds[index] = start_chunk_rand;
-            }
+        advance_m1(start);
```

Comment:
I'm not sure there's any need to advance start by -1 here, as it goes out of scope immediately afterwards.

Reply:
Good catch.